util/chunk/disk.go

```diff
@@ -0,0 +1,284 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package chunk
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+	"sync"
+
+	"github.com/pingcap/parser/terror"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/disk"
+	"github.com/pingcap/tidb/util/stringutil"
+)
+
+const (
+	writeBufSize = 128 * 1024
+	readBufSize  = 4 * 1024
+)
+
+var bufWriterPool = sync.Pool{
+	New: func() interface{} { return bufio.NewWriterSize(nil, writeBufSize) },
+}
+
+var bufReaderPool = sync.Pool{
+	New: func() interface{} { return bufio.NewReaderSize(nil, readBufSize) },
+}
+
+var tmpDir = path.Join(os.TempDir(), "tidb-server-hashJoin")
+
+func init() {
+	_ = os.RemoveAll(tmpDir)
+	_ = os.Mkdir(tmpDir, 0755)
+}
+
+// ListInDisk represents a slice of chunks storing in temporary disk.
+type ListInDisk struct {
+	fieldTypes []*types.FieldType
+	// offsets stores the offsets in disk of all RowPtr,
+	// the offset of one RowPtr is offsets[RowPtr.ChkIdx][RowPtr.RowIdx].
+	offsets [][]int64
+	// offWrite is the current offset for writing
+	offWrite int64
+
+	disk        *os.File
+	bufWriter   *bufio.Writer
+	diskTracker *disk.Tracker // track disk usage.
+}
+
+var chunkListInDiskLabel fmt.Stringer = stringutil.StringerStr("chunk.ListInDisk")
+
+// NewListInDisk creates a new ListInDisk with field types.
+func NewListInDisk(fieldTypes []*types.FieldType) *ListInDisk {
+	l := &ListInDisk{
+		fieldTypes: fieldTypes,
+		// TODO(fengliyuan): set the quota of disk usage.
+		diskTracker: disk.NewTracker(chunkListInDiskLabel, -1),
+	}
+	return l
+}
+
+// GetDiskTracker returns the memory tracker of this List.
+func (l *ListInDisk) GetDiskTracker() *disk.Tracker {
+	return l.diskTracker
+}
+
+// Add adds a chunk to the ListInDisk. Caller must make sure the input chk
+// is not empty and not used any more and has the same field types.
+func (l *ListInDisk) Add(chk *Chunk) (err error) {
+	if chk.NumRows() == 0 {
+		panic("chunk appended to List should have at least 1 row")
```

Comment:
How about returning an error here?

Reply:
Copied from https://github.com/pingcap/tidb/blob/master/util/chunk/list.go#L102-L104.
It works well before, could we create another issue if it is really a problem?

Reply:
How about returning an error here and using another PR to fix https://github.com/pingcap/tidb/blob/e1c6dfa/util/chunk/list.go#L102-L104.
Returning an error in these cases is safer than triggering a panic.

Reply:
Done.